### PR TITLE
[Bugfix] Fix logits processor when prompt_logprobs is not None

### LINKED ISSUE #2800: [bug] AssertionError with prompt_logprobs and logits_processors both set
version 0.3.0
with prompt_logprobs and logits_processors both set
```
def proc(token_ids, logits_row):
    return logits_row
sampling_params = SamplingParams(temperature=0.8, top_p=0.95, prompt_logprobs=100, logits_processors=[proc])
```
got error
```
  File "/opt/conda/lib/python3.10/site-packages/vllm/model_executor/layers/sampler.py", line 74, in forward
    logits = _apply_logits_processors(logits, sampling_metadata)
  File "/opt/conda/lib/python3.10/site-packages/vllm/model_executor/layers/sampler.py", line 162, in _apply_logits_processors
    assert logits_row_idx == logits.shape[0]
AssertionError
```

### PR DESCRIPTION
**Pull request by huyiwen**
This is a resubmission of #3023.

This handles `prompt_logprobs` by skipping rows in logits added for the prompt tokens (prompt logprobs are not processed)

Comparing to the previous one, this pr resolves the [comment](https://github.com/vllm-project/vllm/pull/3023#issuecomment-2040557561) of @tjohnson31415.

Related issue: #2141 
Related PR: #2800
FIX #2800

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/layers/logits_processor.py [hunk 1]
```diff
@@ -84,21 +84,29 @@ def _apply_logits_processors(
...
     for seq_ids, sampling_params in sampling_metadata.seq_groups:
         logits_processors = sampling_params.logits_processors
         if logits_processors:
             found_logits_processors = True
             for seq_id in seq_ids:
+                # handle prompt_logprobs by skipping rows in logits added for
+                # the prompt tokens (prompt logprobs are not processed)
+                if (seq_id < sampling_metadata.num_prompts
+                        and sampling_params.prompt_logprobs is not None):
+                    logits_row_idx += sampling_metadata.prompt_lens[seq_id]
```
Comment (njhill) [inline vllm/model_executor/layers/logits_processor.py:98]:
```suggestion
                    logits_row_idx += sampling_metadata.prompt_lens[seq_id] - 1
```
and revert the other `logits_row_idx` changes? I think that would make things a bit clearer.

File: vllm/model_executor/layers/logits_processor.py [hunk 2]
```diff
@@ -84,21 +84,29 @@ def _apply_logits_processors(
...
+    logits_row_idx = -1
     found_logits_processors = False
     for seq_ids, sampling_params in sampling_metadata.seq_groups:
         logits_processors = sampling_params.logits_processors
         if logits_processors:
             found_logits_processors = True
             for seq_id in seq_ids:
+                # handle prompt_logprobs by skipping rows in logits added for
+                # the prompt tokens (prompt logprobs are not processed)
+                if (seq_id < sampling_metadata.num_prompts
```
Comment (njhill) [inline vllm/model_executor/layers/logits_processor.py:96]: `seq_id`s are allocated using a global counter, so you'll have to use `enumerate` as is done in https://github.com/vllm-project/vllm/pull/3876.

**Comment by njhill** [comment]
@huyiwen any chance you could add a unit test for this too? :pray:

File: vllm/model_executor/layers/logits_processor.py (hunk 2 above)
Comment (huyiwen) [reply vllm/model_executor/layers/logits_processor.py]: I'm not totally sure what happens when `seq_ids` has more than one value. I hope the new fix is on point, especially since it looks like `logits_row_idx` might get added up multiple times for each `seq_id`.

**Review by njhill — APPROVED** [review]
Thanks again @huyiwen!

File: vllm/model_executor/layers/logits_processor.py [hunk 3]
```diff
@@ -86,11 +86,17 @@ def _apply_logits_processors(
...
+        seq_ids, sampling_params = seq_group
         logits_processors = sampling_params.logits_processors
         if logits_processors:
             found_logits_processors = True
             for seq_id in seq_ids:
+                # handle prompt_logprobs by skipping rows in logits added for
+                # the prompt tokens (prompt logprobs are not processed)
+                if (i < sampling_metadata.num_prompts
+                        and sampling_params.prompt_logprobs is not None):
+                    logits_row_idx += sampling_metadata.prompt_lens[i] - 1
```
Comment (njhill) [inline vllm/model_executor/layers/logits_processor.py:99]: Actually @huyiwen I think this should be moved higher, prior to the `if logits_processors:` line.

In the prompt case, `len(seq_ids)` should always be 1 (you could also add an assert like @tjohnson31415 did [here](https://github.com/vllm-project/vllm/pull/3876/files#diff-a49206d5c29b28f53c18c4c18265a89eb06f11102b2f4d6ee95123ea788a602eR122)).

**Review by njhill — CHANGES_REQUESTED** [review]

File: vllm/model_executor/layers/logits_processor.py (hunk 2 above)
Comment (njhill) [reply vllm/model_executor/layers/logits_processor.py]: > it looks like `logits_row_idx` might get added up multiple times for each `seq_id`.

There should only be one `seq_id` in the prompt case. However I think these rows need to be skipped whether or not the request has logits processors (see my comment below).

File: tests/samplers/test_logits_processor.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        logits_processors=[pick_vllm],
+        prompt_logprobs=3,
+        max_tokens=max_tokens,
+    )
+
+    # test prompt_logprobs is not None
+    vllm_model.model._add_request(prompt=example_prompts[0],
+                                  sampling_params=params_with_logprobs,
+                                  prompt_token_ids=None)
+
```
Comment (njhill) [inline tests/samplers/test_logits_processor.py:40]: @huyiwen perhaps add one more request here which requests `prompt_logprobs` but _doesn't_ have any `logits_processors` (that should fail I think without the latest requested change).

Comment (huyiwen) [reply tests/samplers/test_logits_processor.py:52]: It fails indeed

**Review by njhill — APPROVED** [review]
Thanks for the contribution @huyiwen
